README.md
replicateBE
================

<!-- README.md is generated from README.Rmd. Please edit that file -->

## Comparative BA-calculation for the EMA’s Average Bioequivalence with Expanding Limits (ABEL)

**Package offered for Use without any Guarantees and Absolutely No
Warranty. No Liability is accepted for any Loss and Risk to Public
Health Resulting from Use of this R-Code.**

## Introduction

The library provides data sets (internal `.rda` and in CSV-format in
`/extdata/`) which support users in a black-box performance
qualification (PQ) of their software installations. The methods given by
the EMA in [Annex
I](https://www.ema.europa.eu/en/documents/other/31-annex-i-statistical-analysis-methods-compatible-ema-bioequivalence-guideline_en.pdf "EMA/582648/2016, 21 September 2016")
for reference-scaling according to the EMA’s [Guideline on the
Investigation of
Bioequivalence](https://www.ema.europa.eu/en/documents/scientific-guideline/guideline-investigation-bioequivalence-rev1_en.pdf "EMA, January 2010")
are implemented. Potential influence of outliers on the variability of
the reference can be assessed by box plots of studentized and
standardized residuals as suggested at a joint [EGA/EMA
workshop](https://www.medicinesforeurope.com/wp-content/uploads/2016/03/EGA_BEQ_QA_WEB_QA_1_32.pdf "London, June 2010").  
In full replicate designs the variability of test and reference
treatments can be assessed by *s<sub>wT</sub>*/*s<sub>wR</sub>* and the
upper confidence limit of *σ<sub>wT</sub>*/*σ<sub>wR</sub>* (required
for the [WHO’s
approach](https://extranet.who.int/prequal/sites/default/files/documents/AUC_criteria_November2018.pdf "Geneva, November 2018")
for reference-scaling of *AUC*).

### Methods

#### Estimation of *CV<sub>wR</sub>* (and *CV<sub>wT</sub>* in full replicate designs)

Called internally by functions `method.A()` and `method.B()`. A linear
model of log-transformed pharmacokinetic (PK) responses and effects  
    *sequence*, *subject(sequence)*, *period*  
where all effects are fixed (*i.e.*, ANOVA). Estimated by the function
`lm()` of library `stats`.

``` r
modCVwR <- lm(log(PK) ~ sequence + subject%in%sequence + period,
                        data = data[data$treatment == "R", ])
modCVwT <- lm(log(PK) ~ sequence + subject%in%sequence + period,
                        data = data[data$treatment == "T", ])
```

#### Method A

Called by function `method.A()`. A linear model of log-transformed PK
responses and effects  
    *sequence*, *subject(sequence)*, *period*, *treatment*  
where all effects are fixed (*i.e.*, ANOVA). Estimated by the function
`lm()` of library `stats`.

``` r
modA <- lm(log(PK) ~ sequence + subject%in%sequence + period + treatment,
                     data = data)
```

#### Method B

Called by function `method.B()`. A linear model of log-transformed PK
responses and effects  
    *sequence*, *subject(sequence)*, *period*, *treatment*  
where *subject(sequence)* is a random effect and all others are fixed.  
Three options are provided

1.  Estimated by the function `lme()` of library `nlme`. Employs degrees
    of freedom equivalent to SAS’ `DDFM=CONTAIN` and Phoenix WinNonlin’s
    `DF Residual`. Implicitly preferred according to the Q\&A document
    and hence, the default of the function.

<!-- end list -->

``` r
modB <- lme(log(PK) ~ sequence +  period + treatment, random = ~1|subject,
                      data = data)
```

2.  Estimated by the function `lmer()` of library `lmerTest`. Employs
    Satterthwaite’s approximation of the degrees of freedom
    `method.B(..., option = 1)` equivalent to SAS’ `DDFM=SATTERTHWAITE`
    and Phoenix WinNonlin’s `DF Satterthwaite`. This is the only
    available method in SPSS.

<!-- end list -->

``` r
modB <- lmer(log(PK) ~ sequence + period + treatment + (1|subject),
                       data = data)
```

3.  Estimated by the function `lmer()` of library `lmerTest`. Employs
    the Kenward-Roger approximation `method.B(..., option = 3)`. This is
    the only available method in JMP.

<!-- end list -->

``` r
modB <- lmer(log(PK) ~ sequence + period + treatment + (1|subject),
                       data = data)
```

#### Average Bioequivalence

Called by function `ABE()`. The model is identical to
[Method A](#methodA). Conventional BE limits (80.00 – 125.00%) are
employed by default. Tighter limits for narrow therapeutic index drugs
(EMA 90.00 – 111.11%) or wider limits (75.00 – 133.33% for
*C<sub>max</sub>* according to the guideline of the Gulf Cooperation
Council (Bahrain, Kuwait, Oman, Qatar, Saudi Arabia, and the United Arab
Emirates) can be specified.

#### Tested designs

##### Four period (full) replicates

`TRTR | RTRT`  
`TRRT | RTTR`  
`TTRR | RRTT`  
`TRTR | RTRT | TRRT | RTTR` <small>(confounded effects, *not
recommended*)</small>  
`TRRT | RTTR | TTRR | RRTT` <small>(confounded effects, *not
recommended*)</small>

##### Three period (full) replicates

`TRT | RTR`  
`TRR | RTT`

##### Two period (full) replicate

`TR | RT | TT | RR` <small>(Balaam’s design; *not recommended* due to
poor power characteristics)</small>

##### Three period (partial) replicates

`TRR | RTR | RRT`  
`TRR | RTR` <small>(Extra-reference design; biased in the presence of
period effects, *not recommended*)</small>

#### Cross-validation

Details about the reference datasets:

``` r
help("data", package = "replicateBE")
?replicateBE::data
```

Results of the 28 reference datasets agree with ones obtained in SAS
(9.4), Phoenix WinNonlin (6.4 – 8.1), STATISTICA (13), SPSS (22.0),
Stata (15.0), and JMP (10.0.2).

## Examples

  - Evaluation of the internal reference dataset 01 of [Annex
    II](https://www.ema.europa.eu/en/documents/other/31-annex-ii-statistical-analysis-bioequivalence-study-example-data-set_en.pdf "EMA, 21 September 2016")
    by Method A.

<!-- end list -->

``` r
library(replicateBE)
res <- method.A(verbose = TRUE, details = TRUE, print = FALSE,
                data = rds01)

Data set DS01 by Method A 
────────────────────────── 
Analysis of Variance Table

Response: log(PK)
                  Df    Sum Sq   Mean Sq  F value     Pr(>F)    
sequence           1   0.00765 0.0076519  0.04783  0.8270958    
period             3   0.69835 0.2327836  1.45494  0.2278285    
treatment          1   1.76810 1.7680980 11.05095  0.0010405 ** 
sequence:subject  75 214.12956 2.8550608 17.84467 < 2.22e-16 ***
Residuals        217  34.71895 0.1599952                        
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

treatment T – R:
  Estimate Std. Error    t value   Pr(>|t|) 
0.14547400 0.04650870 3.12788000 0.00200215 
217 Residual Degrees of Freedom

cols <- c(12, 15:19) # Relevant columns
print(round(res[cols], 2), row.names=FALSE)

CVwR(%) EL.lo(%) EL.hi(%) CI.lo(%) CI.hi(%)  PE(%)
  46.96    71.23    140.4   107.11   124.89 115.66
```

  - Same dataset evaluated by Method B. Outlier assessment,
    recalculation of *CV<sub>wR</sub>* after exclusion of outliers, new
    expanded limits.

<!-- end list -->

``` r
res <- method.B(ola = TRUE, verbose = TRUE, details = TRUE, print = FALSE,
                data = rds01)

Outlier analysis
 (externally) studentized residuals
 Limits (2×IQR whiskers): -1.717435, 1.877877
 Outliers:
 subject sequence  stud.res
      45     RTRT -6.656940
      52     RTRT  3.453122

 standarized (internally studentized) residuals
 Limits (2×IQR whiskers): -1.69433, 1.845333
 Outliers:
 subject sequence stand.res
      45     RTRT -5.246293
      52     RTRT  3.214663

Data set DS01: Method B by lme (option=2; equivalent to SAS’ DDFM=CONTAIN) 
────────────────────────────────────────────────────────────────────────── 
            numDF denDF       F-value p-value
(Intercept)     1   217 6162.79177905  <.0001
sequence        1    75    0.01402184  0.9061
period          3   217    0.82478701  0.4815
treatment       1   217    9.86464146  0.0019

treatment T – R:
       Value    Std.Error           DF      t-value      p-value 
1.460882e-01 4.651301e-02 2.170000e+02 3.140803e+00 1.919596e-03

cols <- c(25, 28:29, 17:19) # Relevant columns
print(round(res[cols], 2), row.names=FALSE)

CVwR.new(%) EL.new.lo(%) EL.new.hi(%) CI.lo(%) CI.hi(%)  PE(%)
      32.16        78.79       126.93   107.17   124.97 115.73
```

## Installation

Install the released version from CRAN

``` r
install.packages("replicateBE", repos = "https://cloud.r-project.org/")
```

Install the development version from GitHub:

``` r
# install.packages("devtools", repos = "https://cloud.r-project.org/")
devtools::install_github("Helmut01/replicateBE")
```


NEWS
Version 1.0.9
Published on Github 2019-07-22
    - Submitted to CRAN.

Version 1.0.8.9000
Published on Github 2019-07-20
    - importFrom(pbkrtest, getKR) since on win-builder.r-project.org error
      'pbkrtest package required for Kenward-Roger's method' (not locally!)
    - Updated links to EMA's documents (again...) in man-pages. Numbered references.
    - More informative warning in get.data() if na-specifier does not
      match the data in columns PK and/or logPK.
    - Kept column logPK only in rds01 and rds02 because given by the
      EMA. Removed in rds06, rds12, rds14, rds18, rds19, rds20, rds21
      (TRTR|RTRR) and rds03, rds17 (TRT|RTR). Man-pages corrected.
      Reduces the footprint of the library by 5%.
      Depends now on R >= 3.5.0 (since .rda written with 3.6.0).
    - Bug fix: Forces logtrans=TRUE if a user asks for an
      internal dataset which does not have the column logPK.
    - If 'path.in' and/or 'path.out' not given or specified
      folder does not exist: Changed cat() to warning().
    - Cosmetic changes in verbose output of method.A().
    - suppressMessages() no more needed (vignette and testthat).
      Issue orginating in rlang and ggplot2 resolved.
      Both packages don't contain a NEWS file. THX Hadley!
    - stringsAsFactors=FALSE in res-list of all methods.
    - Added Kenward-Roger degrees of freedom to method.B (..., option=3).
    - Removed superfluous options(ow) after models (THX to DL).
    - Minor changes in man-page of method.A and the vignette.

Version 1.0.8
Published on Github 2019-06-14
    - Another fix (THX to Uwe Ligges @CRAN).
      Resubmitted to CRAN.

Version 1.0.7
Published on Github 2019-06-14
    - Fixes suggested by <cran-submissions@r-project.org>
      Resubmitted to CRAN.

Version 1.0.6
Published on Github 2019-06-12
    - Submitted to CRAN (passed checks at R-release and R-devel).

Version 1.0.5.9003
Published on GitHub 2019-06-12
    - Solution: text between \list{}{}
      Discouraged acc. R-Extension Manual 2.1.1!
    - LaTeX problems partly fixed (THX to Duncan Murdoch on r-help).
    - Simplified and reordered sources of man-pages.
    - Updated WHO-reference.
    - Changed {\eqn{foo\textsubscript}{bar}} to simple {\eqn{foo_{bar}}
      in two man pages.
      Created another project where all conditional html/latex-statements
      were removed. Still tons of errors in creating the PDF manual.
    - StartupWarnings seemingly unavoidable until rlang will
      be corrected
    - Small correction in text of box plot of CV.calc.R
      Studentized/standardized does not exactly agree with SAS
      (uses type=2 by default).
    - First version of vignette.
    - Added \tests for testthat().

Version 1.0.5.9002
Published on GitHub 2019-05-30
    - Updated links in man-pages to reflect changes in the FDA's site.
    - Updated links to reference data in info.data.R
    - Updated links in CSV-files.
    - Renamed S3 methods (prefixed with 'repBE') acc. to
      https://github.com/r-lib/rlang/issues/669 THX to MT!
      However, doesn't help. TODO!
    - The warning in CHECK dissappeared in R 3.6.0. Throws other ones
      in examples which change the working directory. All changed to
      \dontrun{}

Version 1.0.5.9001
Published on GitHub 2019-04-23
    - Updated links in man-pages to reflect changes in the EMA's site.

Version 1.0.5.9000
Published on GitHub 2017-11-25
    - TODO: CHECK throws a WARNING.
      checking for unstated dependencies in examples ... WARNING
      Warning in parse(file = files, n = -1L) :
        invalid input found on input connection 'replicateBE-Ex.R'
      parse() is used in the man-pages of method.B and data. Neither
      \dontrun{}, commenting the code out, or completelly removing
      the examples resolves the issue. Suggestions are welcome!
    - Added data set 27 for Balaam's design (TR|RT|TT|RR) and data
      set 28 (TTRR|RRTT); both simulated with homoscedasticy. The
      former only for completeness (poor power). The latter might be
      useful for steady state studies. Man-pages for both.
    - Working on LaTeX syntax for PDF manual. Poor support of UTF-8
      is a joke!
    - Bug fix: Import from XLS with a header.
    - Balaam's design implemented. Not documented yet.
    - Compare variabilites according to the FDA's warfarin guidance.
    - Bug fix: get.data() CSV-file: Issue with characters in the header
      which were == sep resolved. [MT]
    - Bug fix: get.data() Sub/seq in data.frame was empty for complete
      sequences.
    - Added data set 26 (TRTR|RTRT). One of the few where we have results
      for comparison.
    - Added data set 25 (TRTR|RTRT) simulated with heteroscedasticy.
    - Shows a 'graphical' presentation of the CI, PE, and BE-limits in
      the result file.
    - Bug-fix: info.env() gave 'path.out' as 'path.in'.
    - Rewrite of reading from file. The header is automatically
      identified (removed as an argument from the calling functions).
      Subject must no more be the first column. The word 'subject' is
      allowed in the header. [MT]
      TODO: CSV-files - characters in the header which are equal to
      'sep' are lost.
    - Bug-fix: mixed model was given in the result file of method.A().
    - Added data set 24 (TRRT|RTTR|TTRR|RRTT) from the FDA. Subject 16
      completely missing. Adapted get.data() for such a case.
    - Calculate the sw-ratio (maybe useful for the WHO).
    - Bug-fix: nTT was NA in get.data() for TRTR|RTRT. Now design
      (full/replicate) is obtained from info.design().
    - Assessment for NTIDs implemented. ABE() has new arguments 'theta1'
      and 'theta2'. Default to 0.80 and 1.25.
    - Argument 'set' can be an empty string "" to support reading from
      CSV-files. Only required for XLS (the name of the sheet) now.
    - Argument 'fence' for box plots instead of the hard-coded 3 as a
      multiplier for IQR. The default 1.5 in most (all?) software packages

R/method.B.R
#################################################
# EMA's 'Method B' (linear mixed effects)       #
# fixed:  sequence, period, treatment           #
# random: subjects                              #
# option=1: nlme/lme (Satterthwaite's DF)       #
# option=2: lmerTest/lmer (CONTAIN/Residual DF) #
#################################################
method.B <- function(alpha = 0.05, path.in = NULL, path.out = NULL,
                     file, set = "", ext, na = ".", sep = ",",
                     dec = ".", logtrans = TRUE, ola = FALSE,
                     print = TRUE, details = FALSE, verbose = FALSE,
                     ask = FALSE, plot.bxp = FALSE, fence = 2,
                     data = NULL, option = 2) {
  exec <- strftime(Sys.time(), usetz=TRUE)
  ret  <- CV.calc(alpha=alpha, path.in=path.in, path.out=path.out,
                  file=file, set=set, ext=ext, na=na, sep=sep,
                  dec=dec, logtrans=logtrans, ola=ola,
                  print=print, verbose=verbose, ask=ask,
                  plot.bxp=plot.bxp, fence=fence, data=data)
  logtrans <- ret$logtrans
  # Add description of the degrees of freedom to the result file
  if (option == 1) DF.suff <- "Satt"
  if (option == 2) DF.suff <- "GL"
  if (option == 3) DF.suff <- "KR"
  results  <- paste0(ret$res.file, "_MethodB_DF_", DF.suff, ".txt")
  # generate variables based on the attribute
  # 2nd condition: Otherwise, the header from a CSV file will be overwritten
  if (!is.null(data) & missing(ext)) {
    info  <- info.data(data)
    file  <- info$file
    set   <- info$set
    ref   <- info$ref
    descr <- info$descr
    ext   <- ""
  }
  os <- Sys.info()[[1]]   # get OS for line-endings in output (Win: CRLF)
  ow <- options("digits") # save options
  options(digits=12)      # increase digits for anova()
  on.exit(ow)             # ensure that options are reset if an error occurs
  if (option == 2) {      # by nlme (Residual DF)
    options(contrasts=c("contr.treatment", "contr.poly"))
    if (logtrans) {       # use the raw data and log-transform internally
      modB <- lme(log(PK) ~ sequence + period + treatment,
                            random = ~1|subject, na.action=na.omit,
                            data=ret$data)
    } else {              # use the already log-transformed data
      modB <- lme(logPK ~ sequence + period + treatment,
                          random = ~1|subject, na.action=na.omit,
                          data=ret$data)
    }
    EMA.B <- summary(modB)
    PE    <- EMA.B$tTable["treatmentT", "Value"]
    CI    <- exp(as.numeric(intervals(modB, which="fixed",
                          level=1-2*alpha)[[1]]["treatmentT", c(1, 3)]))
    DF    <- EMA.B$tTable["treatmentT", "DF"]
    if (verbose) {
      name <- paste0(file, set)
      cat("\nData set", paste0(name,
          ": Method B by lme (option=2; ",
          "equivalent to SAS\u2019 DDFM=CONTAIN)"),
          paste0("\n", paste0(rep("\u2500", 70+nchar(name)), collapse="")), "\n")
      print(anova(modB), digits=6)
      cat("\ntreatment T \u2013 R:\n")
      print(signif(EMA.B$tTable["treatmentT", ], 7))
      cat("\n")
    }
  } else {              # by lmer/lmerTest (Satterthwaite's or Kenward-Roger DF)
    if (logtrans) {     # use the raw data and log-transform internally
      modB <- lmer(log(PK) ~ sequence + period + treatment + (1|subject),
                             data=ret$data)
    } else {            # use the already log-transformed data
      modB <- lmer(logPK ~ sequence + period + treatment + (1|subject),
                           data=ret$data)
    }
    if (option == 1) {
      EMA.B <- summary(modB, ddf="Satterthwaite")
    } else {
      EMA.B <- summary(modB, ddf="Kenward-Roger")
    }
    PE    <- EMA.B$coefficients["treatmentT", "Estimate"]
    CI    <- exp(PE + c(-1, +1) *
                 qt(1-alpha, EMA.B$coef["treatmentT", "df"]) *
                 EMA.B$coef["treatmentT", "Std. Error"])
    DF    <- EMA.B$coefficients["treatmentT", "df"]
    if (verbose) {
      if (option == 1) {
         cat("\nData set", paste0(file, set,
             ": Method B by lmer (option=1; equivalent to SAS\u2019 DDFM=SATTERTHWAITE)"),
            paste0("\n", paste0(rep("\u2500", 81), collapse="")), "\n")
        print(anova(modB, ddf="Satterthwaite"))
      } else {
        df.txt <- "3; equivalent to SAS\u2019 DDFM=KENWARDROGER)"
        cat("\nData set", paste0(file, set,
            ": Method B by lmer (option=3; equivalent to SAS\u2019 DDFM=KENWARDROGER)"),
            paste0("\n", paste0(rep("\u2500", 80), collapse="")), "\n")
        print(anova(modB, ddf="Kenward-Roger"))
     }
      cat("\ntreatment T \u2013 R:\n")
      print(signif(EMA.B$coefficients["treatmentT", ], 7))
      cat("\n")
    }
  } # end of evaluation by option=2 (lme) or option=1/3 (lmer)
  PE  <- exp(PE)
  res <- data.frame(ret$type, paste0("B-", option), ret$n, ret$nTT, ret$nRR,
                    paste0(ret$Sub.Seq, collapse="|"),
                    paste0(ret$Miss.seq, collapse="|"),
                    paste0(ret$Miss.per, collapse="|"), alpha,
                    DF, ret$CVwT, ret$CVwR, ret$sw.ratio,
                    ret$sw.ratio.upper, ret$BE1, ret$BE2, CI[1], CI[2],
                    PE, "fail", "fail", "fail", log(CI[2])-log(PE),
                    paste0(ret$ol, collapse="|"), ret$CVwR.new,
                    ret$sw.ratio.new, ret$sw.ratio.new.upper, ret$BE.new1,
                    ret$BE.new2, "fail", "fail", "fail",
                    stringsAsFactors=FALSE)
  names(res)<- c("Design", "Method", "n", "nTT", "nRR", "Sub/seq",
                 "Miss/seq", "Miss/per", "alpha", "DF", "CVwT(%)",
                 "CVwR(%)", "sw.ratio", "sw.ratio.CL", "EL.lo(%)",
                 "EL.hi(%)", "CI.lo(%)", "CI.hi(%)", "PE(%)",
                 "CI", "GMR", "BE", "log.half-width", "outlier",
                 "CVwR.new(%)", "sw.ratio.new", "sw.ratio.new.CL",
                 "EL.new.lo(%)", "EL.new.hi(%)",
                 "CI.new", "GMR.new", "BE.new")
  if (ret$BE2 == 1.25) { # change column names if not scaling
    colnames(res)[which(names(res) == "EL.lo(%)")] <- "BE.lo(%)"
    colnames(res)[which(names(res) == "EL.hi(%)")] <- "BE.hi(%)"
  }
  if (!is.na(ret$BE.new2) & ret$BE.new2 == 1.25) { # change column names if not scaling
    colnames(res)[which(names(res) == "EL.new.lo(%)")] <- "BE.new.lo(%)"
    colnames(res)[which(names(res) == "EL.new.hi(%)")] <- "BE.new.hi(%)"
  }
  # Convert CVs, limits, PE, and CI (up to here as fractions) to percent
  res$"CVwT(%)" <- 100*res$"CVwT(%)"
  res$"CVwR(%)" <- 100*res$"CVwR(%)"
  if ("BE.lo(%)" %in% names(res)) { # conventional limits
    res$"BE.lo(%)" <- 100*res$"BE.lo(%)"
    res$"BE.hi(%)" <- 100*res$"BE.hi(%)"
  } else {                          # expanded limits
    res$"EL.lo(%)" <- 100*res$"EL.lo(%)"
    res$"EL.hi(%)" <- 100*res$"EL.hi(%)"
  }
  if (!is.na(res$"CVwR.new(%)")) { # only if recalculated CVwR
    res$"CVwR.new(%)" <- 100*res$"CVwR.new(%)"
    if ("BE.new.lo(%)" %in% names(res)) { # conventional limits
      res$"BE.new.lo(%)" <- 100*res$"BE.new.lo(%)"
      res$"BE.new.hi(%)" <- 100*res$"BE.new.hi(%)"
    } else {                              # expanded limits
      res$"EL.new.lo(%)" <- 100*res$"EL.new.lo(%)"
      res$"EL.new.hi(%)" <- 100*res$"EL.new.hi(%)"
    }
  }
  res$"PE(%)"    <- 100*res$"PE(%)"
  res$"CI.lo(%)" <- 100*res$"CI.lo(%)"
  res$"CI.hi(%)" <- 100*res$"CI.hi(%)"
  if (round(res$"CI.lo(%)", 2) >= 100*ret$BE1 &
      round(res$"CI.hi(%)", 2) <= 100*ret$BE2)
    res$CI <- "pass"  # CI within acceptance range
  if (round(res$"PE(%)", 2) >= 80 & round(res[["PE(%)"]], 2) <= 125)
    res$GMR <- "pass" # PE within 80.00-125.00%
  if (res$CI == "pass" & res$GMR == "pass")
    res$BE <- "pass"  # if passing both, conclude BE
  if (!is.na(res$"CVwR.new(%)")) {
    if (round(res$"CI.lo(%)", 2) >= 100*ret$BE.new1 &
        round(res$"CI.hi(%)", 2) <= 100*ret$BE.new2)
      res$CI.new <- "pass"  # CI within acceptance range
    res$GMR.new <- res$GMR
    if (res$CI.new == "pass" & res$GMR.new == "pass")
      res$BE.new <- "pass"  # if passing both, conclude BE
  }
  if (details) { # results in default (7 digits) precision
    ret <- res
    if (as.character(res$outlier) == "NA") {
      # remove superfluous columns if ola=FALSE or ola=TRUE
      # and no outlier(s) detected
      ret <- ret[ , !names(ret) %in% c("outlier", "CVwR.new(%)",
                                       "sw.ratio.new", "EL.new.lo(%)",
                                       "EL.new.hi(%)", "CI.new",
                                       "GMR.new", "BE.new")]
    }
    #class(ret) <- "repBE"
    return(ret)
  }
  # Round percents to two decimals according to the GL
  res$"CVwT(%)" <- round(res$"CVwT(%)", 2)
  res$"CVwR(%)" <- round(res$"CVwR(%)", 2)
  if ("BE.lo(%)" %in% names(res)) { # conventional limits
    res$"BE.lo(%)" <- round(res$"BE.lo(%)", 2)
    res$"BE.hi(%)" <- round(res$"BE.hi(%)", 2)
  } else {                          # expanded limits
    res$"EL.lo(%)" <- round(res$"EL.lo(%)", 2)
    res$"EL.hi(%)" <- round(res$"EL.hi(%)", 2)
  }
  res$"PE(%)"    <- round(res$"PE(%)", 2)
  res$"CI.lo(%)" <- round(res$"CI.lo(%)", 2)
  res$"CI.hi(%)" <- round(res$"CI.hi(%)", 2)
  if (!is.na(res$"CVwR.new(%)")) { # only if recalculated CVwR
  res$"CVwR.new(%)" <- round(res$"CVwR.new(%)", 2)
    if ("BE.new.lo(%)" %in% names(res)) { # conventional limits
      res$"BE.new.lo(%)" <- round(res$"BE.new.lo(%)", 2)
      res$"BE.new.hi(%)" <- round(res$"BE.new.hi(%)", 2)
    } else {                          # expanded limits
      res$"EL.new.lo(%)" <- round(res$"EL.new.lo(%)", 2)
      res$"EL.new.hi(%)" <- round(res$"EL.new.hi(%)", 2)
    }
  }
  overwrite <- TRUE # default
  if (print) { # to file in UTF-8
    if (ask & file.exists(results)) {
      answer <- tolower(readline("Results already exists. Overwrite the file [y|n]? "))
      if(answer != "y") overwrite <- FALSE
    }
    if (overwrite) { # either the file does not exist or should be overwritten
      # only binary mode supports UTF-8 and different line endings
      res.file <- file(description=results, open="wb")
      res.str  <- info.env(fun="method.B", option=option, path.in, path.out,
                           file, set, ext, exec, data)
      if (os == "Windows") res.str <- gsub("\n", "\r\n", res.str) # CRLF
      if (os == "Darwin")  res.str <- gsub("\n", "\r", res.str)   # CR
      writeBin(charToRaw(res.str), res.file)
      close(res.file)
    }
  }
  # insert DF of treatment difference and alpha into the text generated by CV.calc()
  cut.pos    <- unlist(gregexpr(pattern="Switching CV", ret$txt))
  left.str   <- substr(ret$txt, 1, cut.pos-1)
  right.str  <- substr(ret$txt, cut.pos, nchar(ret$txt))
  insert.str <- "Degrees of freedom : "
  if (option == 1) insert.str <- paste0(insert.str, sprintf("%7.3f (Satterthwaite)", DF))
  if (option == 2) insert.str <- paste0(insert.str, sprintf("%3i", DF))
  if (option == 3) insert.str <- paste0(insert.str, sprintf("%7.3f (Kenward-Roger)", DF))
  insert.str <- paste0(insert.str, "\nalpha              :   ", alpha,
                " (", 100*(1-2*alpha), "% CI)\n")
  txt <- paste0(left.str, insert.str, right.str)
  if (!is.na(res$"CVwR.new(%)")) {
    txt1 <- paste0("\n\nAssessment based on original CVwR",
                   sprintf(" %.2f%%", res$"CVwR(%)"))
    txt <- paste0(txt, txt1, "\n", paste0(rep("\u2500", nchar(txt1)-2), collapse=""))
  }
  txt <- paste0(txt,
                "\nConfidence interval: ", sprintf("%6.2f%% ... %6.2f%%",
                                                   res$"CI.lo(%)", res$"CI.hi(%)"),
                "  ", res$CI,
                "\nPoint estimate     : ", sprintf("%6.2f%%", res$"PE(%)"),
                "              ", res$GMR,
                "\nMixed (CI & PE)    :                      ", res$BE, "\n")
  txt <- paste0(txt,
                repBE.draw.line(called.from="ABEL", L=ret$BE1, U=ret$BE2,
                                lo=CI[1], hi=CI[2], PE=PE), "\n")
  if (!is.na(res$"CVwR.new(%)")) {
    txt1 <- paste0("\nAssessment based on recalculated CVwR",
                   sprintf(" %.2f%%", res$"CVwR.new(%)"))
    txt <- paste0(txt, txt1, "\n", paste0(rep("\u2500", nchar(txt1)-1), collapse=""))
    txt <- paste0(txt, "\nConfidence interval: ", res$CI.new,
                  "\nPoint estimate     : ", res$GMR.new,
                  "\nMixed (CI & PE)    : ", res$BE.new, "\n")
    txt <- paste0(txt,
                  repBE.draw.line(called.from="ABEL", L=ret$BE.new1, U=ret$BE.new2,
                                  lo=CI[1], hi=CI[2], PE=PE), "\n")
  }
  if (res$Design == "TRR|RTR")
    txt <- paste0(txt, "Note: The extra-reference design assumes lacking period effects. ",
                  "The treatment\ncomparison will be biased in the presence of a ",
                  "true period effect.\n")
  if (res$Design %in% c("TRTR|RTRT|TRRT|RTTR", "TRRT|RTTR|TTRR|RRTT"))
    txt <- paste0(txt, "Note: Confounded effects; design not recommended.\n")
  if (print & overwrite) {
    res.file <- file(description=results, open="ab")
    res.str  <- txt # UNIXes LF
    if (os == "Windows") res.str <- gsub("\n", "\r\n", txt) # CRLF
    if (os == "Darwin")  res.str <- gsub("\n", "\r", txt)   # CR
    writeBin(charToRaw(res.str), res.file)
    close(res.file)
  }
} # end of function method.B()
